### PR TITLE
add function splitByRegexp

### PR DESCRIPTION
**Pull request by abel-cheng**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add function splitByRegexp

Detailed description / Documentation draft:

Add function splitByRegexp
https://github.com/ClickHouse/ClickHouse/issues/19386

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Functions/FunctionsStringArray.h [hunk 1]
```diff
@@ -263,6 +263,89 @@ class SplitByStringImpl
...
+    void set(Pos pos_, Pos end_)
+    {
+        pos = pos_;
+        end = end_;
+    }
+
+    /// Get the next token, if any, or return false.
+    bool get(Pos & token_begin, Pos & token_end)
+    {
+        if (!re)
```
Comment (vdimir) [inline src/Functions/FunctionsStringArray.h:317]: Query `SELECT splitByRegexp('x*', 'xxx123xxx')` returns `['','123xxx']`, looks a bit weird. But I'm not sure how it should be. Maybe we can test if empty string match regex and consider it as empty regex. Or maybe it's useless corner-case and current behavior appropriate.

File: tests/queries/0_stateless/01866_split_by_regexp.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+select splitByRegexp('\\d+', 'a1ba5ba8b');
```
Comment (vdimir) [inline tests/queries/0_stateless/01866_split_by_regexp.sql:1]:
```suggestion
select splitByRegexp('\\d+', x) from (select arrayJoin(['a1ba5ba8b', 'a11ba5689ba891011b']) x);
```

File: tests/queries/0_stateless/01866_split_by_regexp.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+select splitByRegexp('', 'abcde');
+select splitByRegexp('<[^<>]*>', x) from (select arrayJoin(['<h1>hello<h2>world</h2></h1>', 'gbye<split>bug']) x);
+select splitByRegexp('ab', '');
+select splitByRegexp('', '');
```
Comment (vdimir) [inline tests/queries/0_stateless/01866_split_by_regexp.sql:5]:
```suggestion
select splitByRegexp('', '');

```

File: docs/en/sql-reference/functions/splitting-merging-functions.md [hunk 4]
```diff
@@ -90,6 +90,53 @@ SELECT splitByString('', 'abcde')
...
+Type: [Array](../../sql-reference/data-types/array.md) of [String](../../sql-reference/data-types/string.md).
+
+-   A non-empty regular expression match occurs at the beginning or end of the string;
+-   There are multiple consecutive non-empty regular expression matches;
+-   The original string `s` is empty while the regular expression is not empty.
+
+**Example**
+
+``` sql
+SELECT splitByRegexp('\\d+', 'a1bc2de3f')
```
Comment (vdimir) [inline docs/en/sql-reference/functions/splitting-merging-functions.md:121]:
```suggestion
SELECT splitByRegexp('\\d+', 'a12bc23de345f')
```

File: docs/en/sql-reference/functions/splitting-merging-functions.md [hunk 5]
```diff
@@ -90,6 +90,53 @@ SELECT splitByString('', 'abcde')
...
+```
+
+``` text
+┌─splitByRegexp('\\d+', 'a1bc2de3f')─┐
+│ ['a','bc','de','f']                │
+└────────────────────────────────────┘
```
Comment (vdimir) [inline docs/en/sql-reference/functions/splitting-merging-functions.md:127]:
```suggestion
┌─splitByRegexp('\\d+', 'a12bc23de345f')─┐
│ ['a','bc','de','f']                    │
└────────────────────────────────────────┘
```

File: docs/en/sql-reference/functions/splitting-merging-functions.md [hunk 6]
```diff
@@ -90,6 +90,53 @@ SELECT splitByString('', 'abcde')
...
+**Returned value(s)**
+
+Returns an array of selected substrings. Empty substrings may be selected when:
+
+Type: [Array](../../sql-reference/data-types/array.md) of [String](../../sql-reference/data-types/string.md).
+
+-   A non-empty regular expression match occurs at the beginning or end of the string;
+-   There are multiple consecutive non-empty regular expression matches;
+-   The original string `s` is empty while the regular expression is not empty.
+
```
Comment (vdimir) [inline docs/en/sql-reference/functions/splitting-merging-functions.md:117]:
```suggestion

-   A non-empty regular expression match occurs at the beginning or end of the string;
-   There are multiple consecutive non-empty regular expression matches;
-   The original string `s` is empty while the regular expression is not empty.

Type: [Array](../../sql-reference/data-types/array.md) of [String](../../sql-reference/data-types/string.md).
```

**Review by vdimir — APPROVED** [review]
Thank you for your PR.
I made minor changes, and left couple of comments.

File: src/Functions/FunctionsStringArray.h (hunk 1 above)
Comment (abel-cheng) [reply src/Functions/FunctionsStringArray.h:317]: This logic is similar with extractAll.
I tried ` SELECT extractAll( 'xxx123xxx', 'x*');` returns `['xxx'] `. These two functions share the same logic.

File: src/Functions/FunctionsStringArray.h [hunk 7]
```diff
@@ -263,6 +263,89 @@ class SplitByStringImpl
...
+
+        if (!col)
+            throw Exception("Illegal column " + arguments[0].column->getName()
+                            + " of first argument of function " + getName() + ". Must be constant string.",
+                            ErrorCodes::ILLEGAL_COLUMN);
+
+        if (!col->getValue<String>().empty())
+            re = Regexps::get<false, false>(col->getValue<String>());
+
+        matches.resize(1);
```
[the remark this answers is not in the document]
Comment (abel-cheng) [reply src/Functions/FunctionsStringArray.h:298]: This is not necessary, I will delete it.

File: tests/queries/0_stateless/01866_split_by_regexp.reference [hunk 8]
```diff
@@ -0,0 +1,6 @@
+['a','ba','ba','b']
```
Comment (vdimir) [inline tests/queries/0_stateless/01866_split_by_regexp.reference:1]:
```suggestion
['a','ba','ba','b']
['a','ba','ba','b']
```

File: src/Functions/FunctionsStringArray.h (hunk 7 above)
Comment (vdimir) [inline src/Functions/FunctionsStringArray.h:298]:
```suggestion
```

File: src/Functions/FunctionsStringArray.h (hunk 1 above)
Comment (vdimir) [reply src/Functions/FunctionsStringArray.h:316]: Ok, let's leave it as is

File: tests/queries/0_stateless/01866_split_by_regexp.reference (hunk 8 above)
Comment (abel-cheng) [reply tests/queries/0_stateless/01866_split_by_regexp.reference:1]: Is this applied? I saw errors in last check.

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-9464.

**Pull request merged**